src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -15,32 +16,71 @@ namespace System.Net.Http.Headers
     // Use HeaderDescriptor.TryGet to resolve an arbitrary header name to a HeaderDescriptor.
     internal readonly struct HeaderDescriptor : IEquatable<HeaderDescriptor>
     {
-        private readonly string _headerName;
-        private readonly KnownHeader? _knownHeader;
+        /// <summary>
+        /// Either a <see cref="KnownHeader"/> or <see cref="string"/>.
+        /// </summary>
+        private readonly object _descriptor;
 
         public HeaderDescriptor(KnownHeader knownHeader)
         {
-            _knownHeader = knownHeader;
-            _headerName = knownHeader.Name;
+            _descriptor = knownHeader;
         }
 
         // This should not be used directly; use static TryGet below
-        internal HeaderDescriptor(string headerName)
+        internal HeaderDescriptor(string headerName, bool customHeader = false)
         {
-            _headerName = headerName;
-            _knownHeader = null;
+            Debug.Assert(customHeader || KnownHeaders.TryGetKnownHeader(headerName) is null, $"The {nameof(KnownHeader)} overload should be used for {headerName}");
+            _descriptor = headerName;
         }
 
-        public string Name => _headerName;
-        public HttpHeaderParser? Parser => _knownHeader?.Parser;
-        public HttpHeaderType HeaderType => _knownHeader == null ? HttpHeaderType.Custom : _knownHeader.HeaderType;
-        public KnownHeader? KnownHeader => _knownHeader;
+        public string Name
+        {
+            get
+            {
+                object descriptor = _descriptor;
+                return descriptor.GetType() == typeof(KnownHeader) ?
+                    Unsafe.As<KnownHeader>(descriptor).Name :
```

Comment:
KnownHeader is sealed; you should get almost the same codegen with an `is` cast and if for some reason you don't, that's something that should be fixable in the JIT. The main difference I'd expect `is` will still incur a null check, but that also means you get the added protection of not getting a null reference exception if default(HeaderDescriptor).Name is used. And with all the headers that are "known", I'd hope the impact of the cast to a sealed string (`descriptor as string`) for the ones that aren't should not be impactful. All this unsafe casting isn't ideal.

What percentage of the improvements you showed are due to the use of Unsafe.As? e.g. if you instead use patterns like:
```C#
public string Name
{
    get
    {
        object descriptor = _descriptor;
        return descriptor is KnownHeader header ? header.Name : descriptor as string;
    }
}
```
what does that do to the numbers?

Reply:
The [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABABgAJiBGAOgCUBXAOwwEt8YaBhCfAB1YAbGFADKIgG6swMXAG4AsAChlrFiKbZB5WNgAmEJoICe5XBigMwGcgAkY+kQBFZYKKz4ZoygN7LyAeR87hLYGDA6DgZGphDAAFYw1uQA+nqu7p7QikqB5P6BxADMlFQUAHLYHAUBfrl5gQDmMBg1DXUNneRxicnpuG4eXlDkALyp/YNZUDldDcQA7OSTmcM0AOItACrGfDAAFACUY+MYuzAQAGb7ANJMEADuTPaOUMcA/G1zeQCqTLjYS6cACCuAAPHdHs8oiIAHz7FZDaCHGiVDjkEBfb4BP4AoE0UFg6ikeGI6aHWZdAC+bRpKnqAWKpQqVRgACY2h0us1Wgz2ljOosJhkkSNWLhyJCni90iMABYwkbvcgK16o1kY4UDVbQch4ZmUzp0vLGwJtJnACAQbQASVwUuhr32EAYNgdMpE5AA1vdpYq0N1XcyVYq0TBDpyBd0EkkbGThmMtVNhoaGqxLuQESLphttucjidyGc9ldbr7HbLDhG+XkudifVCPSNxrjASDwe7FaTs8MKVHfv82zRRF6PDamKwMM6g6rZWG+zWukKLAwYKmTVGYIJcDAo3Xvq38SOxxOpy6bA2/a8F9iArORGHE4f20Syt3taKb7ehZctDv14EpoBEB+R8haVq2va5ZNmy05utBireghrwBuewb3lA86RouAQ9LGyw9rq4xpIRMxRummbxrq4qSshsohtee79oEl4Vp64wYQBnTPsOo58OOk5wQxc6sl+2LLpYa5RiBeRbjuTE4Q0PHHvxp5CaxTZid8GGPuMVEjPqxJcfMSy/tuUmKSBdJ0qo6hQJo2g7loMB6JQbK0Y2iq+OaJTEuQj4+OQPJyOQ1lAA==) is quite a bit worse, but I'll remeasure without `Unsafe.As`.


> that also means you get the added protection of not getting a null reference exception if default(HeaderDescriptor).Name is used

Arguably hitting an exception here is even desired as it indicates our logic is broken / there was concurrent access to the collection that messed things up.

Reply:
> The codegen is quite a bit worse

Not really... and it's way safer and easier to reason about. We should only be playing Unsafe.As tricks when it's absolutely critical. 

As an aside, you should still only read _descriptor once into a local. 

> Arguably hitting an exception here is even desired

Hitting a null reference exception in conjunction with use of Unsafe in situations where concurrent access is possible (even if not supported) is never desirable.

Reply:
Additionally, I would hope that this (and other routines here) are typically getting inlined. That will impact codegen in nontrivial ways.

Making these routines more complicated may actually inhibit inlining, which is very likely counterproductive.

Reply:
With the following change as `pr`: https://github.com/MihaZupan/runtime-1/commit/f67226ad1acda3d02d79f69953a143c5a06da10a

**NonValidatedAdd_NonValidatedEnumerate**

| Toolchain | RequestHeaders |        Mean |     Error |     StdDev |      Median | Ratio | RatioSD |
|---------- |--------------- |------------:|----------:|-----------:|------------:|------:|--------:|
|        pr |              2 |    91.51 ns |  0.139 ns |   0.662 ns |    91.16 ns |  1.00 |    0.00 |
|    unsafe |              2 |    89.91 ns |  0.301 ns |   1.436 ns |    88.97 ns |  0.98 |    0.02 |
|           |                |             |           |            |             |       |         |
|        pr |              4 |   158.69 ns |  0.145 ns |   0.687 ns |   158.38 ns |  1.00 |    0.00 |
|    unsafe |              4 |   160.77 ns |  0.777 ns |   4.049 ns |   161.48 ns |  1.01 |    0.02 |
|           |                |             |           |            |             |       |         |
|        pr |              6 |   296.10 ns |  1.190 ns |   6.203 ns |   292.65 ns |  1.00 |    0.00 |
|    unsafe |              6 |   274.07 ns |  0.972 ns |   4.614 ns |   275.46 ns |  0.93 |    0.02 |
|           |                |             |           |            |             |       |         |
|        pr |              8 |   354.51 ns |  1.504 ns |   7.069 ns |   350.97 ns |  1.00 |    0.00 |
|    unsafe |              8 |   345.02 ns |  1.451 ns |   6.975 ns |   340.95 ns |  0.97 |    0.03 |
|           |                |             |           |            |             |       |         |
|        pr |             12 |   781.55 ns |  2.801 ns |  13.887 ns |   775.41 ns |  1.00 |    0.00 |
|    unsafe |             12 |   729.47 ns |  3.631 ns |  17.520 ns |   729.39 ns |  0.93 |    0.03 |
|           |                |             |           |            |             |       |         |
|        pr |             16 | 1,244.16 ns |  2.477 ns |  11.713 ns | 1,237.19 ns |  1.00 |    0.00 |
|    unsafe |             16 | 1,145.68 ns |  4.649 ns |  22.433 ns | 1,156.79 ns |  0.92 |    0.03 |
|           |                |             |           |            |             |       |         |
|        pr |             32 | 3,936.57 ns |  6.690 ns |  32.917 ns | 3,923.61 ns |  1.00 |    0.00 |
|    unsafe |             32 | 3,587.23 ns | 39.021 ns | 184.894 ns | 3,466.43 ns |  0.91 |    0.05 |

**SendAsync (Create and serialize the request, parse the response and enumerate 4 response headers)**

| Toolchain | RequestHeaders |     Mean |     Error |    StdDev |   Median | Ratio | RatioSD |
|---------- |--------------- |---------:|----------:|----------:|---------:|------:|--------:|
|        pr |              4 | 1.640 μs | 0.0061 μs | 0.0301 μs | 1.639 μs |  1.00 |    0.00 |
|    unsafe |              4 | 1.622 μs | 0.0049 μs | 0.0235 μs | 1.607 μs |  0.99 |    0.02 |
|           |                |          |           |           |          |       |         |
|        pr |              8 | 2.029 μs | 0.0031 μs | 0.0148 μs | 2.032 μs |  1.00 |    0.00 |
|    unsafe |              8 | 2.041 μs | 0.0051 μs | 0.0252 μs | 2.053 μs |  1.01 |    0.01 |
|           |                |          |           |           |          |       |         |
|        pr |             16 | 3.464 μs | 0.0045 μs | 0.0214 μs | 3.463 μs |  1.00 |    0.00 |
|    unsafe |             16 | 3.389 μs | 0.0072 μs | 0.0352 μs | 3.395 μs |  0.98 |    0.01 |

> Making these routines more complicated may actually inhibit inlining, which is very likely counterproductive.

At least for calls like `Equals(HeaderDescriptor)`, they're getting inlined with or without `Unsafe`.

Reply:
The difference is certainly measurable when looking at header operations in isolation, but not game-changing when compared to the total cost of sending a request.

> We should only be playing Unsafe.As tricks when it's absolutely critical.

Not critical, I'm just fond of `Unsafe` :)

Reply:
Removed the Unsafe and minimized changes

Reply:
I've updated the benchmarks in the initial post - the differences on `SendAsync` with/without unsafe are insignificant.